functional/generics-type-classes.md

```diff
@@ -0,0 +1,268 @@
+
+## Advantages
+
+First, it allows fields that are common to multiple states to be de-duplicated.
+By making the non-shared fields generic, they are implemented once.
+
+Second, it makes the `impl` blocks easier to read, because they are broken down
+by state. Methods common to all states are typed once in one block, and methods
+unique to one state are in a separate block.
+
+Both of these mean there is less code and it is better organized.
+
+## Disadvantages
+
+This increases code size. Hopefully the compiler will be more intelligent about
+the way monomorphization is implemented in the future.
+
+## Alternatives
+
+If a type seems to need a "split API" due to construction or partial
+initialization, consider the
+[Builder Pattern](../patterns/creational/builder.md) instead.
+
+If the API between types does not change -- only the behavior does -- then the
+[Strategy Pattern](../patterns/behavioural/strategy.md) is better used instead.
+
+## See also
+
+This pattern is used throughout the standard library:
+
+* `Vec<u8>` can be cast from a String, unlike every other type of `Vec<T>`.[^1]
+* They can also be cast into a binary heap, but only if they contain a type
+  that implements the `Ord` trait.[^2]
+* The `to_string` method was specialized for `Cow` only of type `str`.[^3]
+
+It is also used by several popular crates to allow API flexibility:
+
+* The `embedded-hal` ecosystem used for embedded devices makes extensive use of
+  this pattern. For example, it allows statically verifying the configuration of
+  device registers used to control embedded pins. When a pin is put into a mode,
+  it returns a `Pin<MODE>` struct, whose generic determines the functions
+  usable in that mode, which are not on the `Pin` itself. [^4]
+
+* The `hyper` HTTP client library uses this to expose rich APIs for different
+  pluggable requests. Clients with different connectors have different methods
+  on them as well as different trait implementations, while a core set of
+  methods apply to any connector. [^5]
+
+* The "type state" pattern -- where an object gains and loses API based on an
+  internal state or invariant -- is implemented in Rust using the same basic
+  concept, and a slightly different techinque. [^6]
+
+[^1] See: [impl From\<CString\> for Vec\<u8\>](
+https://doc.rust-lang.org/stable/src/std/ffi/c_str.rs.html#799-801)
+
+
+[^2] See: [impl\<T\> From\<Vec\<T, Global\>\> for BinaryHeap\<T\>](
+https://doc.rust-lang.org/stable/src/alloc/collections/binary_heap.rs.html#1345-1354)
+
+[^3] See: [impl\<'_\> ToString for Cow\<'_, str>](
+https://doc.rust-lang.org/stable/src/alloc/string.rs.html#2235-2240)
+
+[^4] Example:
+[https://docs.rs/stm32f30x-hal/0.1.0/stm32f30x_hal/gpio/gpioa/struct.PA0.html](
+https://docs.rs/stm32f30x-hal/0.1.0/stm32f30x_hal/gpio/gpioa/struct.PA0.html)
+
+[^5] See:
+[https://docs.rs/hyper/0.14.5/hyper/client/struct.Client.html](
+https://docs.rs/hyper/0.14.5/hyper/client/struct.Client.html)
+
+[^6] See:
```

Comment:
```suggestion
[^1]: See: [impl From\<CString\> for Vec\<u8\>](
https://doc.rust-lang.org/stable/src/std/ffi/c_str.rs.html#799-801)


[^2]: See: [impl\<T\> From\<Vec\<T, Global\>\> for BinaryHeap\<T\>](
https://doc.rust-lang.org/stable/src/alloc/collections/binary_heap.rs.html#1345-1354)

[^3]: See: [impl\<'_\> ToString for Cow\<'_, str>](
https://doc.rust-lang.org/stable/src/alloc/string.rs.html#2235-2240)

[^4]: Example:
[https://docs.rs/stm32f30x-hal/0.1.0/stm32f30x_hal/gpio/gpioa/struct.PA0.html](
https://docs.rs/stm32f30x-hal/0.1.0/stm32f30x_hal/gpio/gpioa/struct.PA0.html)

[^5]: See:
[https://docs.rs/hyper/0.14.5/hyper/client/struct.Client.html](
https://docs.rs/hyper/0.14.5/hyper/client/struct.Client.html)

[^6]: See:
```
This should fix footnotes, see https://github.com/rust-lang/mdBook/blob/master/tests/dummy_book/src/first/markdown.md